[GTK][WPE] Merge functions of registering and unregistering script message handler in WebKitUserContentManager API

siwei-li opened the pull request:
#### 0e04927f8eba1d5140e63a9b88970e0b321c701b
<pre>
[GTK][WPE] Merge functions of registering and unregistering script message handler in WebKitUserContentManager API
<a href="https://bugs.webkit.org/show_bug.cgi?id=248748">https://bugs.webkit.org/show_bug.cgi?id=248748</a>

Reviewed by Carlos Garcia Campos.

For the new upcoming API (guarded with ENABLE_2022_GLIB_API),
replace webkit_user_content_manager_register_script_message_handler_in_world and
webkit_user_content_manager_register_script_message_handler
into a single function that always receives a script world
and NULL can be passed to use the default one, along with
their counterpart functions for unregistering the message handler.

* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
(webkit_user_content_manager_unregister_script_message_handler):
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestInputMethodContext.cpp:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestSSL.cpp:
(WebSocketTest::~WebSocketTest):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
(UserScriptMessageTest::unregisterHandler):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp:
(testWebContextSecurityFileXHR):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:

Canonical link: <a href="https://commits.webkit.org/260442@main">https://commits.webkit.org/260442@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c851e2adca358a93d06a08533778635d1c72a791

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108312 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17404 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41176 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117429 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/116811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112198 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/76/builds/18929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8683 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100530 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114081 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/76/builds/18929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97356 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42082 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29000 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10242 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30344 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10982 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7253 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16392 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49941 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12568 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3923 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -471,23 +471,29 @@ gboolean webkit_user_content_manager_register_script_message_handler(WebKitUserC
  * webkit_user_content_manager_unregister_script_message_handler:
  * @manager: A #WebKitUserContentManager
  * @name: Name of the script message channel
+ * @world_name: the name of a #WebKitScriptWorld
```
```suggestion
 * @world_name: (nullable): the name of a #WebKitScriptWorld
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
```diff
@@ -571,6 +577,8 @@ gboolean webkit_user_content_manager_register_script_message_handler_in_world(We
  *
  * Since: 2.22
+ *
+ * Deprecated: 2.40
```
No need to deprecate this:

 * For the new API version, you should just remove it entirely.
 * For the old API version, it's still needed because you cannot change the old webkit_user_content_manager_unregister_script_message_handler(), so it cannot be deprecated.

mcatanzaro reviewed (changes requested):
So we need to support both the old and new API versions in the same codebase. That means you'll need to use `#if ENABLE(2022_GLIB_API)` guards to vary the behavior based on which API is built, being careful not to change the original API.

The doc comments actually have to be split out into separate files so avoid breaking the introspection scanner. Let's call them WebKitUserContentManager1.cpp and WebKitUserContentManager2.cpp. You'll need to add both new files to Sources.txt, then modify PlatformGTK.cmake and PlatformWPE.cmake to conditionally add one file or the other to `WEBKITGTK_SOURCES_FOR_INTROSPECTION` and `WPE_SOURCES_FOR_INTROSPECTION`. There are examples of how to do this in https://github.com/WebKit/WebKit/commit/da67032774ee7d7d35ad2746bb374d2f1ef2fc0d and https://github.com/WebKit/WebKit/commit/4b24459ebb4af8ba7b6547991d14194c4f13de31.

If you have questions, don't hesitate to ask!

mcatanzaro commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestInputMethodContext.cpp:
```diff
@@ -281,7 +281,7 @@ class InputMethodTest: public WebViewTest {
     ~InputMethodTest()
     {
-        webkit_user_content_manager_unregister_script_message_handler(m_userContentManager.get(), "imEvent");
+        webkit_user_content_manager_unregister_script_message_handler(m_userContentManager.get(), "imEvent", "");
```
Are you sure this is correct? I thought using an empty string as the world name produced a new script world that is distinct from the default script world. Shouldn't you pass `NULL` here instead?

siwei-li replied:
Right NULL should be passed here, thank you for pointing this out

mcatanzaro replied:
(Oops, I really meant `nullptr`. We don't use `NULL` anywhere in C++ code anymore.)

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager2.cpp:
```diff
@@ -0,0 +1,61 @@
+ *
+ * See also webkit_user_content_manager_register_script_message_handler().
+ *
+ * Since: 2.8
```
Remove this or use since: 2.40

carlosgcampos reviewed (commented):
I don't like doc comments in the headers, but in this case, since the headers are generated it would be simpler to add the docs there only for these functions and we don't need to add new source files. What do you think @mcatanzaro ?

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.h.in:
```diff
@@ -80,7 +80,12 @@ webkit_user_content_manager_register_script_message_handler            (WebKitUs
                                                                         const gchar              *name);
+#else
+                                                                        const gchar               *name,
+                                                                        const gchar               *world_name);
```
You have one extra space here:

```suggestion
                                                                        const gchar              *name,
                                                                        const gchar              *world_name);
```

carlosgcampos reviewed (commented):
Great! This look much simpler. But I think for the new API we should also remove the register in world variant. If we merge the unregister we should do the same fo the register ones.

mcatanzaro reviewed (commented):
Nice! Don't forget to update the title of the commit since you're no longer just touching the unregister functions.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestConsoleMessage.cpp:
```diff
@@ -63,7 +63,11 @@ class ConsoleMessageTest : public WebViewTest {
         webkit_user_content_manager_register_script_message_handler(m_userContentManager.get(), "console");
+#else
+        webkit_user_content_manager_register_script_message_handler_with_reply(m_userContentManager.get(), "console", nullptr);
+#endif
```
We don't want to remove `webkit_user_content_manager_register_script_message_handler` from new api, we want to remove the in_world version adding the world nam parameter to `webkit_user_content_manager_register_script_message_handler`

siwei-li committed:
[GTK][WPE] Merge functions of registering and unregistering script message handler in WebKitUserContentManager API
https://bugs.webkit.org/show_bug.cgi?id=248748

Reviewed by Carlos Garcia Campos.

For the new upcoming API (guarded with ENABLE_2022_GLIB_API),
replace webkit_user_content_manager_register_script_message_handler_in_world and
webkit_user_content_manager_register_script_message_handler
into a single function that always receives a script world
and NULL can be passed to use the default one, along with
their counterpart functions for unregistering the message handler.

* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.cpp:
(webkit_user_content_manager_unregister_script_message_handler):
* Source/WebKit/UIProcess/API/glib/WebKitUserContentManager.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestInputMethodContext.cpp:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestSSL.cpp:
(WebSocketTest::~WebSocketTest):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitUserContentManager.cpp:
(UserScriptMessageTest::unregisterHandler):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebContext.cpp:
(testWebContextSecurityFileXHR):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:

Canonical link: https://commits.webkit.org/260442@main
